llama-cpp-2/src/context/kv_cache.rs

```diff
@@ -121,6 +121,19 @@ impl LlamaContext<'_> {
         unsafe { llama_cpp_sys_2::llama_kv_cache_seq_pos_max(self.context.as_ptr(), seq_id) }
     }
 
+    /// Remove all tokens within the specified range `[p0, p1)` from the KV cache
+    /// Returns `false` only when partial sequence removals fail. Full sequence removals always succeed.
+    ///
+    /// # Parameters
+    ///
+    /// * `seq_id` - The sequence id to remove the tokens from. If < 0, matches all sequences
+    /// * `p0` - The start position of the cache to remove. If < 0, the entire cache is removed up to `p1`
+    /// * `p1` - The end position of the cache to remove. If < 0, the entire cache is removed starting from `p0`
+    #[must_use]
+    pub fn kv_cache_seq_rm(&mut self, seq_id: i32, p0: i32, p1: i32) -> bool {
```

Comment:
I would prefer for these (unsure how consistently this is done elsewhere) that we don't have special behaviour for negative ints (however well documented it is). An Option<NonZeroU16> is more clear.

Reply:
that's a great call!

would you like me to apply this change to the `seq_id` as well? it shares the same "-1 means all" semantics, tho I see in other kv cache-related methods so far we've left that an `i32`

---

by the way, i don't have a use case for it today, but there are already models with context lengths greater than 65k, so `u16` for `llama_pos` can't fully express the range of valid values for such large models.. but of course jumping to a `u32` introduces the overflow problem on the downcast to `i32` for values between 2B & 4B, so it's probably a fine tradeoff to opt for simplicity & leave as `u16`s for now.

alternatively, we could accept `u32` for `llama_pos` & return a result if `i32::try_from` fails, rather than a bool, which might be a bit more rust-y anyhow?

Reply:
for example:
```rust
    /// Remove all tokens within the specified range `[p0, p1)` from the KV cache
    /// Returns `false` only when partial sequence removals fail. Full sequence removals always succeed.
    ///
    /// # Returns
    /// A `Result` indicating whether the operation was successful. If the sequence id or position
    /// exceed the maximum i32 value, no remonval is attempted and an `Err` is returned.
    ///
    /// # Parameters
    ///
    /// * `seq_id` - The sequence id to remove the tokens from. If `None`, matches all sequences
    /// * `p0` - The start position of the cache to remove. If `None`, the entire cache is removed up to `p1`
    /// * `p1` - The end position of the cache to remove. If `None`, the entire cache is removed starting from `p0`
    #[must_use]
    pub fn kv_cache_seq_rm(
        &mut self,
        seq_id: Option<u32>,
        p0: Option<u32>,
        p1: Option<u32>,
    ) -> Result<bool, TryFromIntError> {
        let seq_id = seq_id.map_or(Ok(-1), i32::try_from)?;
        let p0 = p0.map_or(Ok(-1), i32::try_from)?;
        let p1 = p1.map_or(Ok(-1), i32::try_from)?;

        Ok(
            unsafe {
                llama_cpp_sys_2::llama_kv_cache_seq_rm(self.context.as_ptr(), seq_id, p0, p1)
            },
        )
    }
```

Reply:
that looks great.

Reply:
and feel free to make similar changes elsewhere.

Reply:
great! do you mind bringing `std::num::TryFromIntError` into the crate's public API, or would you prefer something using `this_error`, like maybe:
```rust
/// Errors that can occur when attempting to manipulate the kv cache
#[derive(Debug, Eq, PartialEq, thiserror::Error)]
pub enum KvCacheConversionError {
    /// Sequence id conversion to i32 failed
    #[error("Provided sequence id is too large for a i32")]
    SeqIdTooLarge(#[source] TryFromIntError),
    /// Position 0 conversion to i32 failed
    #[error("Provided start position is too large for a i32")]
    P0TooLarge(#[source] TryFromIntError),
    /// Position 1 conversion to i32 failed
    #[error("Provided end position is too large for a i32")]
    P1TooLarge(#[source] TryFromIntError),
}
```

Reply:
ok, [ran with something like this](https://github.com/utilityai/llama-cpp-rs/pull/510/commits/7d1b2d53e2b677420539a339d4e9b522a206427b), but let me know if you'd prefer something else!